File: array_section.py
from typing import List


def reverse_str(str: str) -> str | None:
  """Function to reverse a string"""

  if not str or len(str) < 2:
    print("Not a valid string")
    return None
  else:
    list = [i for i in str]
    list.reverse()
    return "".join(list)


def reverse_str2(str: str) -> str | None:
  """Function to reverse a string"""

  if not str or len(str) < 2:
    print("Not a valid string")
    return None
  else:
    str_length = len(str)
    reverse_str = []
    for i in range(str_length - 1, -1, -1):
      reverse_str.append(str[i])

    return "".join(reverse_str)


def reverse_str3(string: str) -> str | None:
  """Function to reverse a string"""

  if not str or not isinstance(string, str) or len(string) < 2:
    print("Not a valid string")
    return None
  else:
    return string[::-1]


def merge_sorted_arrays(list_a: List = [], list_b: List = []) -> List | None:
  """Function that takes in two arrays and merges them"""

  if not list_a or not list_b:
    if not list_a and list_b:
      list_b.sort()
      return list_b
    elif not list_b and list_a:
      list_a.sort()
      return list_a
    else:
      return []
  else:
    list_a.extend(list_b)
    list_a.sort()
    return list_a


def check_index_exists(list, index):
  """Check if index exists"""

  try:
    list[index]
  except IndexError:
    return False
  return True


def merge_sorted_arrays2(list_a: List = [], list_b: List = []) -> List | None:
  """Function that takes in two arrays and merges them"""

  if not list_a or not list_b:
    if not list_a and list_b:
      list_b.sort()
      return list_b
    elif not list_b and list_a:
      list_a.sort()
      return list_a
    else:
      return []
  else:
    sorted_array = []
    list_a.sort()
    list_b.sort()
    i = 0
    j = 0
    while i < len(list_a) or j < len(list_b):
      if (check_index_exists(list_a, i)
          and not check_index_exists(list_b, j)) or (check_index_exists(
              list_a, i) and list_a[i] < list_b[j]):
        sorted_array.append(list_a[i])
        i += 1
      else:
        sorted_array.append(list_b[j])
        j += 1
    return sorted_array


if __name__ == "__main__":
  print(reverse_str("Nakul is cool"))
  print(reverse_str2("Nakul is cool"))
  print(reverse_str3("Nakul is cool"))
  print(reverse_str3(123))
  print(merge_sorted_arrays([1, 8, 0], [2, 7, 6]))
  print(merge_sorted_arrays2([1, 8, 0], [2, 7, 6]))

File: arraySection.js
function reverse(str) {
  if (!str || typeof str !== "string" || str.length < 2) {
    console.log(`Please provide a valid string as ${str} is not valid`);
    return null;
  } else {
    const strSplit = str.split("");
    var reverseStr = "";
    for (let i = 0; i < strSplit.length; i++) {
      reverseStr = reverseStr + strSplit[strSplit.length - 1 - i];
    }
    return reverseStr;
  }
}

console.log("My name is Nakul");
console.log(reverse("My name is Nakul"));
console.log(reverse(1));
console.log(reverse("N"));

function reverse2(str) {
  if (!str || typeof str !== "string" || str.length < 2) {
    console.log(`Please provide a valid string as ${str} is not valid`);
    return null;
  } else {
    const stringLength = str.length;
    let totalItems = [];
    for (i = stringLength; i >= 0; i--) {
      totalItems.push(str[i - 1]);
    }
    return totalItems.join("");
  }
}

function reverse3(str) {
  return str.split("").reverse().join("");
}

const reverse4 = (str) => str.split("").reverse().join("");
const reverse5 = (str) => [...str].reverse().join("");

console.log(reverse2("My name is Nakul"));
console.log(reverse3("My name is Nakul"));
console.log(reverse4("My name is Nakul"));
console.log(reverse5("My name is Nakul"));
console.log(reverse2(1));
console.log(reverse2(""));

function mergeSortedArrays(arr1, arr2) {
  if (typeof arr1 !== "object" && typeof arr2 !== "object") {
    console.log("Error");
    return null;
  } else {
    if (!arr1 && arr2) {
      return arr2.sort();
    } else if (arr1 && !arr2) {
      return arr1.sort();
    } else {
      const sortedArray = [];
      arr1.sort();
      arr2.sort();
      let item1 = arr1[0];
      let item2 = arr2[0];
      let i = 1;
      let j = 1;
      while (item1 || item2) {
        console.log(item1, item2, i, j);
        if (!item2 || item1 <= item2) {
          sortedArray.push(item1);
          item1 = arr1[i];
          i++;
        } else {
          sortedArray.push(item2);
          item2 = arr2[j];
          j++;
        }
      }
      return sortedArray;
    }
  }
}

function mergeSortedArrays2(arr1, arr2) {
  return arr1.concat(arr2);
}

console.log(mergeSortedArrays([1, 8, 0], [2, 7, 6]));
console.log(mergeSortedArrays2([1, 2], ["nakul", "is"]));
